src/signum/signal.clj

```
@@ -0,0 +1,103 @@
+;;   Copyright (c) 7theta. All rights reserved.
+;;   The use and distribution terms for this software are covered by the
+;;   MIT License (https://opensource.org/licenses/MIT) which can also be
+;;   found in the LICENSE file at the root of this distribution.
+;;
+;;   By using this software in any fashion, you are agreeing to be bound by
+;;   the terms of this license.
+;;   You must not remove this notice, or any others, from this software.
+
+(ns signum.signal
+  (:require [clojure.core.async :refer [chan sliding-buffer close!
+                                        go-loop >!! alts!]]
+            [utilis.exception :refer [throw-if]]
+            [utilis.fn :refer [fsafe]]
+            [utilis.string :as ust])
+  (:import [clojure.lang IRef IDeref IObj IMeta]))
+
+(def ^:dynamic *tracker* nil)
+
+(declare pr-signal)
+
+(deftype Signal [backend watches meta-map]
+
+  IDeref
+  (deref [this]
+    ((fsafe *tracker*) :deref this)
+    (throw-if (or (agent-error backend) (clojure.core/deref backend))))
+
+  IRef
+  (addWatch [this watch-key watch-fn]
+    (when (empty? @watches)
+      (add-watch backend :signum/signal
+                 (fn [_key _ref _old-value new-value]
+                   (doseq [{:keys [value-ch]} (vals @watches)]
+                     (>!! value-ch [new-value])))))
+    (let [stop-ch (chan)
+          value-ch (chan (sliding-buffer 1))]
+      (swap! watches assoc watch-key
+             {:stop-ch stop-ch
+              :value-ch value-ch
+              :value-loop (go-loop [old-value @backend]
+                            (let [[[value] channel] (alts! [stop-ch value-ch])]
+                              (when (not= channel stop-ch)
+                                (watch-fn watch-key this old-value value)
```

Is there any chance that watch-fn might throw? The go block will swallow the exception.

---

Good catch. If the watch function throws we should dispose the subscription. Need to figure out how to notify the subscriber that we punted them.

---

Is this maybe where we would look into something like `https://github.com/IGJoshua/farolero` ?

---

Actually never mind, it requires a thread context and we're in core-async land.